broker-core/src/main/java/io/zeebe/broker/workflow/processor/deployment/DeploymentCreatedProcessor.java

```diff
@@ -55,32 +57,50 @@ public void processRecord(
     }
 
     for (final Workflow workflowRecord : deploymentEvent.workflows()) {
-      if (workflowRecord.getVersion() != 1) {
-        closeExistingMessageStartEventSubscriptions(workflowState, workflowRecord, streamWriter);
+      if (!isLatestWorkflow(workflowRecord)) {
+        continue;
       }
+
+      closeExistingMessageStartEventSubscriptions(workflowState, workflowRecord, streamWriter);
       openMessageStartEventSubscriptions(workflowState, workflowRecord, streamWriter);
     }
   }
 
+  private boolean isLatestWorkflow(Workflow workflow) {
+    return workflowState
+            .getLatestWorkflowVersionByProcessId(workflow.getBpmnProcessId())
+            .getVersion()
+        == workflow.getVersion();
+  }
+
   private void closeExistingMessageStartEventSubscriptions(
       WorkflowState workflowState, Workflow workflowRecord, TypedStreamWriter streamWriter) {
-    final int previousVersion = workflowRecord.getVersion() - 1;
-    final DeployedWorkflow previousWorkflow =
-        workflowState.getWorkflowByProcessIdAndVersion(
-            workflowRecord.getBpmnProcessId(), previousVersion);
-    if (previousWorkflow == null
-        || previousWorkflow.getWorkflow().getStartEvents().stream().noneMatch(e -> e.isMessage())) {
+    final DeployedWorkflow lastMsgWorkflow = findLastMessageStartWorkflow(workflowRecord);
+    if (lastMsgWorkflow == null) {
       return;
     }
 
-    final long previousWorkflowKey = previousWorkflow.getKey();
-    final MessageStartEventSubscriptionRecord subscriptionRecord =
-        new MessageStartEventSubscriptionRecord();
-    subscriptionRecord.setWorkflowKey(previousWorkflowKey);
-
+    subscriptionRecord.reset();
+    subscriptionRecord.setWorkflowKey(lastMsgWorkflow.getKey());
     streamWriter.appendNewCommand(MessageStartEventSubscriptionIntent.CLOSE, subscriptionRecord);
   }
 
+  private DeployedWorkflow findLastMessageStartWorkflow(final Workflow workflowRecord) {
+    DeployedWorkflow lastMsgWorkflow;
```

Comment:
I think the declaration here is unnecessary, since you do not use it afterwards.